src/syntaxtree.c

```diff
@@ -549,7 +549,7 @@ static Expr SyntaxTreeCodeValue(Obj node)
     UInt1 tnum = GetTypeTNum(node);
     Obj   value = ElmRecST(tnum, node, "value");
     Expr  expr = NewStatOrExpr(tnum, sizeof(UInt), 0);
-    Int   ix = AddValueToBody(value);
+    Int   ix = AddValueToBody(CopyObj(value, 0));
```

Comment:
I am really torn on this whole concept. On the one hand, it is quite powerful and the implementation is very easy (thanks to the recent addition of `AddValueToBody` and the whole code driving it).
But at the same time, it's a major departure from how things worked so far. Previously, you could print a function, and that gave you code text that could be read in to essentially reproduce the function body. This is now now the case anymore. Perhaps that's not a problem, though (and strictly speaking, this already could happen if you printed a nested function containing hvar references; but there, at least you could have printed the parent function to "fix" it.

Now, I could provide as value, say, an immutable string. Printing the resulting function would give something that indistinguishable from the same function with a string expression, but they'd behave differently. Or, even worse, I could use a *mutable* string. Then it'd look the same, but would work radically differently. E.g the function might modify that string value, and after each call to it, printing the function could show something else.

Don't get me wrong, I am not intending to shoot this down, but as always I'd like us to think through some of the long-term consequences first... and perhaps come up with some ideas how they might be addressable. That is also tightly related to the question how to make these "values" accessible without using the syntax tree (de)coding.

One idea for that would be to add some new syntax to mark an expression, which says: "when coding, evaluate this expression immediately and insert a reference to the resulting value". So one could write something very roughly in the vein of this (before @sebasguts gets me wrong on this, too: Of course I am not saying these things must be implemented before this PR can be merged; what I want is that we at least verify that we see a way forward, before possibly ourselves into a corner!):
```
gap> f := x -> x + EvalImmediately( 2^30 );;
gap> Print(f);
function ( x )
    return x + EvalImmediately( 1073741824 );
end
gap> g := function(x) local y; y := EvalImmediately( [] ); Add(y, x); return y; end;;
gap> Display(g);
function ( x )
    local y;
    y := EvalImmediately( [  ] );
    Add( y, x );
    return y;
end
gap> g(1);
[ 1 ]
gap> g(2);
[ 1, 2 ]
gap> Display(g);
function ( x )
    local y;
    y := EvalImmediately( [ 1, 2 ] );
    Add( y, x );
    return y;
end
```
To clarify how this is meant to work: `EvalImmediately` here would be a syntactic element, not a function, i.e., similar to how `IsBound` and `Unbind` work. It would only be valid in coding mode (though I guess we could allow it in interpreted mode, as a no-op). When reading `EvalImmediately( expr )`, then `expr` is evaluated immediately when the final `)` is read; thus no references to hvars (arguments or local variables) of the surrounding functions are possible, as those are not bound at parsing time. Once the expression is evaluated, it is inserted as a `EXPR_VALUE` into the function body. And printing an `EXPR_VALUE` would be done by printing `EvalImmediately( VALUE )` instead of just `VALUE`. 

I think that's relatively straight forward to explain and implement. But of course it's still far from perfect, i.e., it does not capture the full power of `EXPR_VALUE` -- so we may want to make it even more general, or restrict what we allow as `EXPR_VALUE` ...
For example, it cannot reflect mutability:
```
gap> h1 := {} -> EvalImmediately( MakeImmutable("string") );;
gap> h2 := {} -> EvalImmediately( "string" );;
gap> Display(h1);
function (  )
    return EvalImmediately( "string" );
end
gap> Display(h2);
function (  )
    return EvalImmediately( "string" );
end
```
So `h1` and `h2` are printed identically, and both return always the same strings; but in one case its mutable, in the other immutable. Oh, and of course, I can do this: `MakeImmutable(h2());` and afterwards the string returned by `h2` is immutable, too...

The specific problem of immutability could be addressed, of course, in various ways (say by adding a special case to the printing code to output `EvalImmediately( MakeImmutable( VALUE ) )` instead of just `EvalImmediately( VALUE )`; but of course that's not enough as there are many objects where the printed form is not enough to reconstruct the object.

BTW, the syntax tree (de)coding also cannot capture the full power of this concept due to the `CopyObj` call you inserted above; and also printing and re-reading the resulting text in general won't produce identical results. Here is an example for that:
```
gap> list := [];
[  ]
gap> func := function(x) Add( EvalImmediately( list ), x ); return EvalImmediately( list ); end;;
gap> Display(func);
function ( x )
    Add( EvalImmediately( [ ] ), x );
    return EvalImmediately( list );
end
gap> func := function(x) Add( EvalImmediately( list ), x ); return EvalImmediately( list ); end;;
gap> func(1);
[ 1 ]
gap> list;
[ 1 ]
gap> Display(func);
function ( x )
    Add( EvalImmediately( list ), x );
    return EvalImmediately( list );
end
```
That kind of code is at the same time pretty cool and pretty horrific (in terms of long-term maintainability). Perhaps one should prevent it, by explicitly specifying that `EvalImmediately` inserts a *copy* of the evaluated object?

Reply:
You are really right about this concerns. Let me address some of them with comments (as I have absolutely no solution, just additional comments)

* For clarification: You do not want to make the functions readable from their display string using this `EvalImmediately`, right? You just want to indicate that this value is somehow special. Because, if you want to make it fully evaluateable again, one needs to store the input string, too, and make it completely static (i.e., not depending on anything global). Furthermore, when working with Syntax Trees, one might not want to even create such an input string, as it might be between hard and up to impossible. However, Just printing the value would certainly be fine for me.

* About the copying: We can just drop the copying when coding the syntax tree. I did not had any special thoughts in mind when doing so, so not making copies at all is not a problem.

* For copying semantics of `EvalImmediately`: We could probably make it produce a `ShallowCopy`, since it already looks like a function call anyway. One probably does not want to have a `DeepCopy` in there always, since this could just be added manually if somebody really wants to copy the value, but if it is the default, it again removes possibilities.